File: python/balance_parens_spec.py
from balance_parens import balance_parens

# Add more test cases!...
print(balance_parens("abc(d)e(fgh))(i)j)k") == "abc(d)e(fgh)(i)jk")
print(balance_parens("abc((d)e(fgh)(i)j(k") == "abc(d)e(fgh)(i)jk")
print(balance_parens("()") == "()")
print(balance_parens("a(b)c)") == "a(b)c")
print(balance_parens("(a)(bdd)c)") == "(a)(bdd)c")
print(balance_parens("a(dbvb)c)") == "a(dbvb)c")
print(balance_parens("a(b)(c)())") == "a(b)(c)()")
print(balance_parens(")(") == "")
print(balance_parens("(((((") == "")
print(balance_parens(")(())(") == "(())")
print(balance_parens("(()()(") == "()()")
print(balance_parens(")())(()()(") == "()()()")

# Challenge: nested parentheses...
print(balance_parens("abc(d)(ef(g(h))ij)k)lm()o)p") == "abc(d)(ef(g(h))ij)klm()op")

File: js/balanceParensSpec.js
let bp = require("./balanceParens");

console.log(balanceParens("abc(d)e(fgh))(i)j)k") === "abc(d)e(fgh)(i)jk")
console.log(balanceParens("abc((d)e(fgh)(i)j(k") === "abc(d)e(fgh)(i)jk")
console.log(balanceParens("()") === "()")
console.log(balanceParens("a(b)c)") === "a(b)c")
console.log(balanceParens("(a)(bdd)c)") === "(a)(bdd)c")
console.log(balanceParens("a(dbvb)c)") === "a(dbvb)c")
console.log(balanceParens("a(b)(c)())") === "a(b)(c)()")
console.log(balanceParens(")(") === "")
console.log(balanceParens("(((((") === "")
console.log(balanceParens(")(())(") === "(())")
console.log(balanceParens("(()()(") === "()()")
console.log(balanceParens(")())(()()(") === "()()()")

console.log(balanceParens("abc(d)(ef(g(h))ij)k)lm()o)p") === "abc(d)(ef(g(h))ij)klm()op")